strcspn: Undefined behavior due to invalid pointer arithmetic

The `strcspn` function in `src/string/strcspn.c` contains a line of code (introduced in commit eb00919a88) that has undefined behavior due to pointer arithmetic going out of bounds:

```
  return (uintptr_t)(__strchrnul(s, *c) - (uintptr_t)a);
```

I believe the intent here is to convert both pointers to `uintptr_t` and then subtract the resulting integers.  But the way the parentheses are placed, it instead converts only `a` to an integer, then subtracts that from the pointer returned by `__strchrnul`.  Since `(uintptr_t)a` will in general be large, the resulting pointer is well outside the range of the object pointed to by `s`, which results in undefined behavior (per C11 6.5.6/8).

In practice, this has no ill effect on any C implementation I'm familiar with, but should be fixed anyway.  I think the following is what was intended:

```
  return (uintptr_t)__strchrnul(s, *c) - (uintptr_t)a;
```

I agree that your change is more readable, but I do not think it actually has any real impact on the operation at all (and Clang agrees). I also don't see where the UB is at in this operation.

- A `uintptr_t` is sized to be an integer that can hold the value of a pointer, and the value held by `a` will not by any larger than what can be held in a `char *`. 
- Since `a` is simply the start address of `s`, and the value returned by `__strchrnul` is an address that is an offset of `s` (that is greater than or equal to `s`/`a`), the resulting value is not outside the range of the string object pointed to by `s`.
- The standard states that the result of pointer math with an integer type is a pointer, and since the value returned by `__strchrnul` is a `char*`, it will be an equivalent mathematical operation as a cast to `uintptr_t` first (would not be true of other pointer types).
- The reason for the cast placement as it was is the original clang warning (reported below for reference) is actually complaining about the implicit conversion from a pointer (`char *`) to an integral type (`size_t` here... who knows why I cast to a `uintptr_t` instead of a `size_t` back then, but that particular implicit conversion works out fine).

Code:

```
return __strchrnul(s, *c) - (uintptr_t)a;
```

Warning:

```
[1/8] Compiling C object src/libc_native.a.p/string_strcspn.c.o
../src/string/strcspn.c:15:10: warning: incompatible pointer to integer conversion returning 'char *' from a function with result type 'size_t' (aka 'unsigned long') [-Wint-conversion]
                return __strchrnul(s, *c) - (uintptr_t)a;
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
[2/8] Compiling C object src/libc.a.p/string_strcspn.c.o
../src/string/strcspn.c:15:10: warning: incompatible pointer to integer conversion returning 'char *' from a function with result type 'size_t' (aka 'unsigned long') [-Wint-conversion]
                return __strchrnul(s, *c) - (uintptr_t)a;
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
[3/8] Compiling C object src/libc_hosted.a.p/string_strcspn.c.o
../src/string/strcspn.c:15:10: warning: incompatible pointer to integer conversion returning 'char *' from a function with result type 'size_t' (aka 'unsigned long') [-Wint-conversion]
                return __strchrnul(s, *c) - (uintptr_t)a;
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
[8/8] Linking target test/libc_test
```

--- a/src/string/__strerror.h
+++ b/src/string/__strerror.h
@@ -102,4 +102,4 @@ E(ESTALE, "Stale file handle")
 E(EDQUOT, "Quota exceeded")
 E(EMULTIHOP, "Multihop attempted")
 
-E(0, "No error")
+E(0, "Success")

--- a/src/string/strcspn.c
+++ b/src/string/strcspn.c
@@ -12,7 +12,7 @@ size_t strcspn(const char* s, const char* c)
 
 	if(!c[0] || !c[1])
 	{
-		return (uintptr_t)(__strchrnul(s, *c) - (uintptr_t)a);
+		return (uintptr_t)(__strchrnul(s, *c) - a);
 	}
 
 	memset(byteset, 0, sizeof byteset);

--- a/src/string/strerror.c
+++ b/src/string/strerror.c
@@ -14,6 +14,8 @@ static const char errmsg[] =
 #include "__strerror.h"
 	;
 
+static const char* invalid_err = "No error information available.";
+
 char* strerror(int err_no)
 {
 	const char* s = NULL;
@@ -23,11 +25,19 @@ char* strerror(int err_no)
 	{
 	}
 
-	for(s = errmsg; i; s++, i--)
+	if(errid[i] != err_no)
 	{
-		for(; *s; s++)
+		s = invalid_err;
+	}
+	else
+	{
+		for(s = errmsg; i; s++, i--)
 		{
+			for(; *s; s++)
+			{
+			}
 		}
 	}
+
 	return (char*)(uintptr_t)s;
 }

